make-release.sh

```diff
@@ -0,0 +1,164 @@
+#!/bin/bash
+# Release process automation script.
+# Used to create branch/tag and update VERSION files
+# and and trigger release by force pushing changes to the release branch
+
+# set to 1 to actually trigger changes in the release branch
+TRIGGER_RELEASE=0
+NOCOMMIT=0
+MAJOR_VERSION_BUMP=0
+
+usage ()
+{
+  echo "Usage: $0 --repo [GIT REPO TO EDIT] --version [VERSION TO RELEASE] [OPTIONS]"
+  echo "Options:"
+  echo "    --help, -h"
+  echo "        Show this message."
+  echo "    --no-commit, -n"
+  echo "        Do not commit changes."
+  echo "    --trigger-release, -t"
+  echo "        Push changes to remote repo, triggering release."
+  echo "Example: $0 --repo git@github.com:eclipse/che-subproject --version 'v3.2.1' --trigger-release"; echo
+}
+
+while [[ "$#" -gt 0 ]]; do
+  case $1 in
+    '-r'|'--repo') REPO="$2"; shift 1;;
+    '-v'|'--version') VERSION="$2"; shift 1;;
+    '-t'|'--trigger-release')
+      if [[ ! ${NOCOMMIT} -eq 0 ]]; then
+        echo "Only one of --trigger-release and --no-commit may be specified"
+        exit 1
+      fi
+      TRIGGER_RELEASE=1; shift 0;;
+    '-n'|'--no-commit')
+      if [[ ! ${NOCOMMIT} -eq 0 ]]; then
+        echo "Only one of --trigger-release and --no-commit may be specified"
+        exit 1
+      fi
+      NOCOMMIT=1; shift 0;;
+    '-h'|'--help'|*) usage; exit 0;;
+  esac
+  shift 1
+done
+
+if [[ ! ${VERSION} ]] || [[ ! ${REPO} ]]; then
+  usage
+  exit 1
+fi
+
+if [[ ${VERSION} != "v"* ]]; then
+  echo "Version must be prefixed with 'v'"
+  exit 1
+fi
+
+CURRENT_MAJOR_VERSION=$(cut -f 1 -d '.' VERSION)
+NEW_MAJOR_VERSION=$(echo "$VERSION" | cut -f 1 -d '.')
+if [[ "${NEW_MAJOR_VERSION}" != "${CURRENT_MAJOR_VERSION}" ]]; then
+  MAJOR_VERSION_BUMP=1
+fi
+
+# derive branch from version
+BRANCH=${VERSION%.*}.x
+
+# if doing a .0 release, use master; if doing a .z release, use $BRANCH
+if [[ ${VERSION} == *".0" ]]; then
+  BASEBRANCH="master"
+else
+  BASEBRANCH="${BRANCH}"
+fi
+
+# work in tmp dir
+TMP=$(mktemp -d); pushd "$TMP" > /dev/null || exit 1
+
+# get sources from ${BASEBRANCH} branch
+echo "Check out ${REPO} to ${TMP}/${REPO##*/}"
+git clone "${REPO}" -q
+cd "${REPO##*/}" || exit 1
+git fetch origin "${BASEBRANCH}":"${BASEBRANCH}"
+git checkout "${BASEBRANCH}"
+
+# create new branch off ${BASEBRANCH} (or check out latest commits if branch already exists), then push to origin
+if [[ "${BASEBRANCH}" != "${BRANCH}" ]]; then
+  git branch "${BRANCH}" || git checkout "${BRANCH}" && git pull origin "${BRANCH}"
+  git push origin "${BRANCH}"
+  git fetch origin "${BRANCH}:${BRANCH}"
+  git checkout "${BRANCH}"
+fi
+
+# change VERSION file
+echo "${VERSION}" > VERSION
+
+# commit change into branch
+if [[ ${NOCOMMIT} -eq 0 ]]; then
+  COMMIT_MSG="[release] Bump to ${VERSION} in ${BRANCH}"
+  git commit -s -m "${COMMIT_MSG}" VERSION
+  git pull origin "${BRANCH}"
```

Comment:
If I understand correctly - it's here to update the branch before pushing and it can produce merge commit appearing.
I believe it should be stable enough if we move pulling before committing command.
Then we higher chance to get our git push command failed, but the value of this increment like up to second. And as small profit out history will be liner.
WDYT?

Reply:
:+1: good idea.